transport/http/jsonrpc/server.go

```diff
@@ -0,0 +1,205 @@
+	return func(s *Server) { s.finalizer = f }
+}
+
+// ServeHTTP implements http.Handler.
+func (s Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		io.WriteString(w, "405 must POST\n")
+		return
+	}
+	ctx := r.Context()
+
+	if s.finalizer != nil {
+		iw := &interceptingWriter{w, http.StatusOK}
+		defer func() { s.finalizer(ctx, iw.code, r) }()
+		w = iw
+	}
+
+	for _, f := range s.before {
+		ctx = f(ctx, r)
+	}
+
+	// Decode the body into an  object
+	var req Request
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		rpcerr := parseError("JSON could not be decoded: " + err.Error())
+		s.logger.Log("err", rpcerr)
+		s.errorEncoder(ctx, rpcerr, w)
+		return
+	}
+
+	// Get the endpoint and codecs from the map using the method
+	// defined in the JSON  object
+	ecm, ok := s.ecm[req.Method]
+	if !ok {
+		err := methodNotFoundError(fmt.Sprintf("Method %s was not found.", req.Method))
+		s.logger.Log("err", err)
+		s.errorEncoder(ctx, err, w)
+		return
+	}
+
+	// Decode the JSON "params"
+	reqParams, err := ecm.Decode(ctx, req.Params)
+	if err != nil {
+		s.logger.Log("err", err)
+		s.errorEncoder(ctx, err, w)
+		return
+	}
+
+	// Call the Endpoint with the params
+	response, err := ecm.Endpoint(ctx, reqParams)
+	if err != nil {
+		s.logger.Log("err", err)
+		s.errorEncoder(ctx, err, w)
+		return
+	}
+
+	for _, f := range s.after {
+		ctx = f(ctx, w)
+	}
+
+	res := Response{
+		JSONRPC: Version,
+	}
+
+	// Encode the response from the Endpoint
+	resParams, err := ecm.Encode(ctx, response)
+	if err != nil {
+		s.logger.Log("err", err)
+		s.errorEncoder(ctx, err, w)
+		return
+	}
+
+	res.Result = resParams
+
+	w.Header().Set("Content-Type", ContentType)
+	json.NewEncoder(w).Encode(res)
+}
+
+// DefaultErrorEncoder writes the error to the ResponseWriter,
+// as a json-rpc error response, with an InternalError status code.
+// The Error() string of the error will be used as the response error message.
+// If the error implements ErrorCoder, the provided code will be set on the
+// response error.
+// If the error implements Headerer, the given headers will be set.
+func DefaultErrorEncoder(_ context.Context, err error, w http.ResponseWriter) {
+	w.Header().Set("Content-Type", ContentType)
+	if headerer, ok := err.(httptransport.Headerer); ok {
+		for k := range headerer.Headers() {
+			w.Header().Set(k, headerer.Headers().Get(k))
+		}
+	}
+
+	e := Error{
+		Code:    InternalError,
+		Message: err.Error(),
+	}
+	if sc, ok := err.(ErrorCoder); ok {
+		e.Code = sc.ErrorCode()
+	}
+
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(Response{
+		JSONRPC: Version,
+		Error:   &e,
+	})
+}
+
+// ErrorCoder is checked by DefaultErrorEncoder. If an error value implements
+// ErrorCoder, the Error will be used when encoding the error. By default,
+// InternalError (-32603) is used.
```

Comment:
I understand the difference between this and StatusCoder, but it took me a second. Maybe it's worth being a bit more explicit, e.g. "...the integer value will be used as the JSONRPC error code"?